rand_distr: Fix dirichlet sample method for small alpha.

WarrenWeckesser opened the pull request:
Generating Dirichlet samples using the method based on samples from the gamma distribution can result in samples being nan if all the values in alpha are sufficiently small.  The fix is to instead use the method based on the marginal distributions being the beta distribution (i.e. the "stick breaking" method) when all values in alpha are small.

More details:

Here's an example where the current method produces `nan`s:

```
use rand::distributions::Distribution;
use rand_distr::Dirichlet;

fn main() {
    let n = 1000;
    println!("Checking {} Dirichlet samples", n);

    let dirichlet = Dirichlet::new(&[0.001, 0.001, 0.001]).unwrap();
    let mut r = rand::thread_rng();
    let mut nancount = 0;
    for _ in 0..n {
        let sample: Vec<f64> = dirichlet.sample(&mut r);
        if sample.iter().any(|x| x.is_nan()) {
            nancount += 1;
        }
    }
    println!("nancount: {}", nancount);
}
```

Typical output:

```
Checking 1000 Dirichlet samples
nancount: 128
```

The method based on samples from the gamma distribution is described in the [wikipedia page on the Dirichlet distribution](https://en.wikipedia.org/wiki/Dirichlet_distribution#Computational_methods).  The problem is that when alpha is small, there is a high probability that the gamma random variate will be 0 (given that we are limited to finite precision floating point).  In fact, we can predict the above result: with alpha=0.001, the probability that a gamma variate G(alpha, 1) will be less than 2.225e-308 (approximately the smallest normal 64 bit floating point value) is approximately 0.4927.  For the Dirichlet distribution with vector parameter [0.001, 0.001, 0.001], the problem occurs when all three gamma variates are 0, which has probabilty (0.4927)**3 = 0.1196.  So when generating 1000 samples, we expect about 120 to contain nan.

A way to avoid this problem is to switch to the less efficient method based on the marginal beta distributions of the Dirichlet distribution.  This method is also described on the [wikipedia page](https://en.wikipedia.org/wiki/Dirichlet_distribution#Marginal_beta_distributions).  In this PR, this method is used when all the alpha values are less than 0.1. This threshold was discussed in the [NumPy PR](https://github.com/numpy/numpy/pull/14924), where it seemed like a reasonable compromise between (i) using the gamma variate method for as wide a range as possible and (ii) ensuring that the probability of generating nans is negligibly small.

WarrenWeckesser commented:
FYI: A similar change was made to numpy in https://github.com/numpy/numpy/pull/14924

dhardy commented:
Thanks for the PR. At a glance this looks good but I'd prefer it have a proper review from someone besides the author (doesn't have to be myself or an existing contributor, if anyone is interested); I didn't find the time yet myself.

WarrenWeckesser commented:
For anyone who might review this: I added some more details to the description above.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -123,16 +123,56 @@ where
+                .zip(self.alpha.iter())
+                .zip(alpha_sum_rl.iter())
+            {
+                let beta = Beta::new(a, b).unwrap();
```
Ideally, this should be done once in `Dirichlet::new`, unless it would require too much storage.

WarrenWeckesser replied:
Done.

vks reviewed (changes requested):
I suggest to move the initialization of the new algorithm for small alpha from `sample` to `new`. This probably requires making `Dirichlet` an enum similar to this:

```rust
struct DirichletFromGamma<F> { alpha: Box<[F]> }
struct DirichletFromBeta<F> { alpha: Box<[F]>, alpha_sum_r1: Box<[F]> }

pub enum Dirichlet<F> where ... {
    FromGamma(DirichletFromGamma<F>), FromBeta(DirichletFromBeta<F>)
}
```

Alternatively, we could use this opportunity to refactor that even more initialization can be moved out of `sample` by switching to something akin to the following:

```rust
struct DirichletFromGamma<F> { samplers: Box<[Gamma<F>]> }
struct DirichletFromBeta<F> { samplers: Box<[Beta<F>]> }
```

vks commented:
The last suggestion however would increase storage by a factor of ca. 4, so it would have to be justified with benchmarks.

WarrenWeckesser commented:
@vks, thanks for the review.  I haven't forgotten about this! A couple other projects moved to the top of my to-do list, but I will get back to it.

dhardy commented:
@WarrenWeckesser can I remind you of this PR?

WarrenWeckesser commented:
@dhardy, yes you can!  It has been slowly bubbling back up my "to do" list.  I'll get back to it this week.

WarrenWeckesser commented:
I pushed an update that implements my interpretation of the comments made by @vks last year.  There are now two structs that implement the two different methods, and `Dirichlet` is an enum with one variant for each type of struct.  `Dirichlet::new()` returns the variant that is appropriate for the values in the input `alpha`.  The initialization of the underlying data that is needed in the `sample()` method of each struct has been moved to the `new()` method of the struct.

This might not be what was intended, and I suspect my Rust code is not idiomatic in a lot of places, so I still view this as a rough draft that will need more iteration.

WarrenWeckesser commented:
I implemented `Dirichlet` as an `enum`, and implemented `new()`, `new_with_size()` and `sample()` for the `enum`.  Over in [`gamma.rs`](https://github.com/rust-random/rand/blob/master/rand_distr/src/gamma.rs), I see that `Gamma` and `ChiSquared` have a pattern similar to `Dirichlet`, where the sampling algorithm depends on the values of the parameters.  Each of those distributions is implemented as a `struct` containing a single field called `repr`, and it is `repr` that is the `enum`, rather than the distribution itself.  Would that be preferable for `Dirichlet`?

WarrenWeckesser commented:
> ... and it is `repr` that is the `enum`, rather than the distribution itself. Would that be preferable for `Dirichlet`?

Answering my own question: yes, I think making `Dirichlet` a `struct` with a field that is an `enum` *is* preferable.  I'll push an update soon.

WarrenWeckesser commented:
I push an update to make the implementation following the style (more or less) of `Gamma` and `ChiSquared`.

WarrenWeckesser commented:
I see that https://github.com/rust-random/rand/pull/1292 is also updating the Dirichlet distribution. This PR will almost certainly have conflicts with https://github.com/rust-random/rand/pull/1292, but I can deal with that if/when https://github.com/rust-random/rand/pull/1292 is merged.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    // Construct a new `Dirichlet` with the given alpha parameter `alpha`.
```
```suggestion
    // Construct a new `Dirichlet` with the given parameter `alpha`.
```

WarrenWeckesser replied:
Done.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+        }
+        let invacc = F::one() / sum;
+        for s in samples.iter_mut() {
+            *s = (*s) * invacc;
```
Similarly:

```suggestion
            *s *= invacc;
```

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+                Some(*sum)
+            })
+            .collect();
+        alpha_sum_rl.reverse();
```
This looks like we could avoid one pass over `alpha` by assigning the values in a loop in the correct order.

WarrenWeckesser replied:
Done.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+        let beta_dists = alpha
+            .iter()
+            .zip(alpha_sum_rl.iter())
+            .map(|t| Beta::new(*t.0, *t.1).unwrap())
```
```suggestion
            .map(|(&alpha, &beta)| Beta::new(alpha, beta).unwrap())
```

WarrenWeckesser replied:
Done (as part of bigger changes).

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+    fn new(alpha: &[F]) -> DirichletFromGamma<F> {
+        let gamma_dists = alpha
+            .iter()
+            .map(|a| Gamma::new(*a, F::one()).unwrap())
```
```suggestion
            .map(|&a| Gamma::new(a, F::one()).unwrap())
```

WarrenWeckesser replied:
Done (as part of bigger changes).

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+    fn new(alpha: &[F]) -> DirichletFromGamma<F> {
+        let gamma_dists = alpha
+            .iter()
+            .map(|a| Gamma::new(*a, F::one()).unwrap())
```
The `unwrap` is a bit problematic here, we probably want to propagate the error as a `Result` instead.

WarrenWeckesser replied:
Done.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+        for (s, beta) in samples.iter_mut().zip(self.samplers.iter()) {
+            let beta_sample = beta.sample(rng);
+            *s = acc * beta_sample;
+            acc = acc * (F::one() - beta_sample);
```
```suggestion
            acc *= F::one() - beta_sample;
```

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -93,12 +241,21 @@ where
 
-        Ok(Dirichlet { alpha: alpha.to_vec().into_boxed_slice() })
+        if alpha.iter().all(|x| *x <= NumCast::from(0.1).unwrap()) {
+            // All the values in alpha are less than 0.1.
```
This comment is a bit redundant to the code in line 244. Maybe we can instead say how 0.1 was chosen as a cutoff?

WarrenWeckesser replied:
Done.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -107,9 +264,7 @@ where
-        Ok(Dirichlet {
-            alpha: vec![alpha; size].into_boxed_slice(),
-        })
+        Ok(Dirichlet::new(vec![alpha; size].as_slice()).unwrap())
```
Can we propagate the error instead of unwrapping?

WarrenWeckesser replied:
Gone -- `new_with_size()` was removed.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -121,27 +276,44 @@ where
 mod test {
     use super::*;
 
+    //
+    // Check that the means of the components of n samples from
+    // the Dirichlet distribution agree with the expected means
+    // with a relative tolerance of rtol.
+    //
+    // This is a crude statistical test, but it will catch egregious
+    // mistakes.  It will also also fail if any samples contain nan.
+    //
```
```suggestion
    /// Check that the means of the components of `n` samples from
    /// the Dirichlet distribution agree with the expected means
    /// with a relative tolerance of `rtol`.
    ///
    /// This is a crude statistical test, but it will catch egregious
    /// mistakes.  It will also also fail if any samples contain `nan`.
```

WarrenWeckesser replied:
Done.

WarrenWeckesser commented:
Thanks @vks.  I'm working on updating this PR with the recent changes from https://github.com/rust-random/rand/pull/1292.  I'll probably incorporate the small changes that you have suggested (e.g. use of in-place operators) while doing that, and then I'll fix the error handling.

WarrenWeckesser commented:
Design question: how do deal with a limitation of const generics?

My update to this PR will have the same basic design, but the use of, for example,`samplers: Box<[Gamma<F>]>` in the `DirichletFromGamma` struct becomes `samplers: [Gamma<F>; N]`.  The problem is that the same field in the `DirichletFromBeta` struct requires only `N - 1` instances of `Beta`.  I can't write `samplers: [Beta<F>; N - 1]` because the expression `N - 1` is not currently supported.  Some work-arounds:
* The compiler tells me

  ```
   = help: const parameters may only be used as standalone arguments, i.e. `N`
   = help: use `#![feature(generic_const_exprs)]` to allow generic const expressions
  ```
  Would it be OK to use that feature?
* Use a length `N` array, i.e. `samplers: [Beta<F>; N]`, and fill in the last element of the array with an unused `Beta` instance. (This is not a very appealing option!)
* Go back to `Box<[Beta<F>]>` for this struct.

Any recommendations?  Any other ideas?

WarrenWeckesser commented:
I'm going to go with `Box<[Beta<F>]>` for now.

dhardy commented:
> Would it be OK to use that feature?

Absolutely not by default. It's an unstable feature and only available in nightly Rust. (In theory you could, behind a `#[cfg(feature = "nightly")]`, but then you need to write and test two implementations. Once the Rust feature is stabilised the next `rand_distr` release can depend on it.)

Rust features can take a long time to make it into stable. I guess we'll have `generic_const_exprs` within a couple of years, but don't know.

> Use a length `N` array

A few dozen wasted bytes, but it's fine.

> Go back to `Box<[Beta<F>]>` for this struct.

Also fine. A little less memory usage, but also less data locality.

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -123,16 +123,56 @@ where
+                    Some(*sum)
+                })
+                .collect();
+            alpha_sum_rl.reverse();
```
Done.

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -123,16 +123,56 @@ where
+        } else {
+            let mut sum = F::zero();
+            for (s, &a) in samples.iter_mut().zip(self.alpha.iter()) {
+                let g = Gamma::new(a, F::one()).unwrap();
```
Done.

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+    // Construct a new `Dirichlet` with the given alpha parameter `alpha`.
+    //
+    // This function is part of a private implementation detail.
+    // It assumes that the input is correct, so no validation is done.
```
Done.

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+        let mut sum = F::zero();
+        for (s, g) in samples.iter_mut().zip(self.samplers.iter()) {
+            *s = g.sample(rng);
+            sum = sum + (*s);
```
`F` doesn't include the assignment operators.

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+        let beta_dists = alpha
+            .iter()
+            .zip(alpha_sum_rl.iter())
+            .map(|t| Beta::new(*t.0, *t.1).unwrap())
```
Done.

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    /// Dirichlet distribution that generates samples using the gamma distribution.
```
Done (although my original intent was to refer to the general idea of the gamma distribution, and not the rust struct `Gamma`, and in that use, "gamma" would not be capitalized).

WarrenWeckesser commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,11 +9,160 @@
+    /// Dirichlet distribution that generates samples using the gamma distribution.
+    FromGamma(DirichletFromGamma<F>),
+
+    /// Dirichlet distribution that generates samples using the beta distribution.
```
Done.

WarrenWeckesser commented:
I have updated the pull request with significant changes, both to handle `Dirichlet` now having a const generic parameter, and to propagate errors from calling `Gamma::new` and `Beta::new`.

In `DirichletFromGamma::new(alpha)`, my rust-fu was not strong enough to figure out how to loop through `alpha` to generate the `Gamma` instances as an array while also handling the possibility of an error from `Gamma::new()`.  So the `samplers` field of `DirichletFromGamma` is now `Box<[Gamma<F>]>` instead of `[Gamma<F>; N]`.

There are now more `Error` variants returned by `Dirichlet::new`, because of more careful validation of `alpha` (e.g. reject subnormal values) and to handle possible errors from `Gamma::new` or `Beta::new`.

dhardy commented:
> my rust-fu was not strong enough to figure out how to loop through alpha to generate the Gamma instances as an array while also handling the possibility of an error from Gamma::new().

The only options I can think of are panic with `catch_unwind` or unsafe code (e.g. use `MaybeUninit` and cast). Rust doesn't have a good option here that I am aware of so keeping your boxed array is probably the best choice.

vks commented:
In such cases, the easiest way is to initialize the array with some default values. We could use `[Option<Gamma<F>>; N]` and initialize it with `[None; N]`, which can then be overwritten iteratively with `Gamma::new()`.

Avoiding the indirection requires unsafe code, or a safe abstraction like [`arrayvec`](https://crates.io/crates/arrayvec).

WarrenWeckesser commented:
> We could use [Option<Gamma<F>>; N] ...

Then when `samplers` is used in `sample()`, instead of

    *s = g.sample(rng);

we would need something like

    *s = g.expect("This should not happen!").sample(rng)

correct?  `new()` must never return without completely initializing `samplers`, so having `samplers` be an array of `Option` smells like an implementation detail of `new()` that shouldn't "leak" out to other code.  But maybe that's fine--it is only a couple lines of code in a private struct.

Another less-than-ideal alternative is to initialize the array as `[Gamma::new(1, 1); N]` to keep the compiler happy, and then overwrite all the values in the array.  Actually, it would look more like

    [Gamma::new(F::one(), F::one()).map_err(|_| DirichletFromGammaError::GammmaNewFailed)?; N]


Yet another alternative for using an array is to assemble the `Gamma` instances in a vector, and then cast the vector to an array with `.try_into()`.  To make this work, I had to add the bound `core::fmt::Debug` to `F`.  Would anyone be interested in seeing this implemented in the pull request?  I don't mind trying out alternatives, if y'all have the patience to review them.

Or, I'm fine with "Just leave it as it is" or "Just do X".

dhardy commented:
> Yet another alternative for using an array is to assemble the Gamma instances in a vector, and then cast the vector to an array with .try_into(). To make this work, I had to add the bound core::fmt::Debug to F.

I'd forgotten (or missed) this method of constructing an array: https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#impl-TryFrom%3CVec%3CT%2C%20A%3E%3E-for-%5BT%3B%20N%5D
I don't see any `Debug` requirement. Looks viable to me?

WarrenWeckesser commented:
Of course, just minutes after I comment, I dig a bit more and realize the cause of the issue. It is `.unwrap()` that has the `Debug` bound.  If I deal with the `Result` returned by `.try_into()` differently, then `Debug` for `F` is not necessary.

vks commented on rand_distr/src/dirichlet.rs:
```diff
@@ -9,12 +9,182 @@
+
+    /// gamma_dists.try_into() failed (in theory, this should not happen).
+    GammaArrayCreationFailed,
+}
```
I'm not 100% sure about the error types here. Ideally, they should refer to invalid parameters supplied by the caller.

However, in this case they are only used internally, so I think this is fine.

vks reviewed (approved):
Sorry for taking a while to come back to this. It looks great, thanks!
